tests/integration/cli/src/object_file_engine_test_c_source.c

```diff
@@ -0,0 +1,145 @@
+#include <stdlib.h>
+
+void wasmer_function__1(void);
+void wasmer_trampoline_function_call__1(void*, void*, void*);
+
+#ifdef __cplusplus
+}
+#endif
+
+void print_wasmer_error()
+{
+    int error_len = wasmer_last_error_length();
+    printf("Error len: `%d`\n", error_len);
+    char* error_str = (char*) malloc(error_len);
+    wasmer_last_error_message(error_str, error_len);
+    printf("Error str: `%s`\n", error_str);
+}
+
+
+int main() {
+        printf("Initializing...\n");
+        wasm_config_t* config = wasm_config_new();
+        wasm_config_set_compiler(config, CRANELIFT);
+        wasm_config_set_engine(config, OBJECT_FILE);
+        wasm_engine_t* engine = wasm_engine_new_with_config(config);
+        wasm_store_t* store = wasm_store_new(engine);
+
+        char* byte_ptr = (char*)&WASMER_METADATA[0];
+
+        // We need to pass all the bytes as one big buffer so we have to do all this logic to memcpy
+        // the various pieces together from the generated header file.
+        //
+        // We should provide a `deseralize_vectored` function to avoid requiring this extra work.
+
+        size_t num_function_pointers
+                = sizeof(function_pointers) / sizeof(void*);
+        size_t num_function_trampolines
+                = sizeof(function_trampolines) / sizeof(void*);
+        size_t num_dynamic_function_trampoline_pointers
+                = sizeof(dynamic_function_trampoline_pointers) / sizeof(void*);
+
+
+        size_t buffer_size = module_bytes_len
+                + sizeof(size_t) + sizeof(function_pointers)
+                + sizeof(size_t) + sizeof(function_trampolines)
+                + sizeof(size_t) + sizeof(dynamic_function_trampoline_pointers);
+
+        char* memory_buffer = (char*) malloc(buffer_size);
+        size_t current_offset = 0;
+        printf("Buffer size: %zu\n", buffer_size);
+
+        memcpy(memory_buffer + current_offset, byte_ptr, module_bytes_len);
+        current_offset += module_bytes_len;
+
+        memcpy(memory_buffer + current_offset, (void*)&num_function_pointers, sizeof(size_t));
+        current_offset += sizeof(size_t);
+
+        memcpy(memory_buffer + current_offset, (void*)&function_pointers[0], sizeof(function_pointers));
+        current_offset += sizeof(function_pointers);
+
+        memcpy(memory_buffer + current_offset, (void*)&num_function_trampolines, sizeof(size_t));
+        current_offset += sizeof(size_t);
+
+        memcpy(memory_buffer + current_offset, (void*)&function_trampolines[0], sizeof(function_trampolines));
+        current_offset += sizeof(function_trampolines);
+
+        memcpy(memory_buffer + current_offset, (void*)&num_dynamic_function_trampoline_pointers, sizeof(size_t));
+        current_offset += sizeof(size_t);
+
+        memcpy(memory_buffer + current_offset, (void*)&dynamic_function_trampoline_pointers[0], sizeof(dynamic_function_trampoline_pointers));
+        current_offset += sizeof(dynamic_function_trampoline_pointers);
```

Comment:
Why do we need this?

Reply:
>         // We need to pass all the bytes as one big buffer so we have to do all this logic to memcpy
>        // the various pieces together from the generated header file.

Deserializing requires one contiguous array of bytes.  We have to pass this info to the ObjectFileEngine for it to deserialize.  We could make a `deserialize_vectored` function which doesn't require memcpying, it allows the bytes to come from multiple sources.